[Gekidou MM-47655] Implement Add Members screen - (merge to main)

jfrerich opened the pull request:
#### Summary
This PR implements the `Add Members` Screen.  It is available by clicking the `Add People` box which can be located via the following ways:

 1) Channel > `...` in header 
 2) Channel > tap the channel header

 This PR does not implement the bottom selected users panel at this time. It is in a separate PR. https://github.com/mattermost/mattermost-mobile/pull/6699

Todo: 
- [ ] guest cannot add people, do we hide the add people box from them?
- [x] return to the Channel Info screen after adding Members
- [x] add toast after adding members
- [x] after removing a user, update the members list to reflect the removed user
- [x] update profile bottom sheet - ticket created https://mattermost.atlassian.net/browse/MM-48470
- [x] fix warning that occurs after selecting a user profile
- [x] needs 'no more users to add to the channel' empty results page
- [x] tutorial should highlight the first row (verify in tablet and phone).  Matt says this was working before so might be introducing a regression.
- [ ] need to pop to correct screen in both cases where add people box can be accessed (look at `Start Call` box for example)
  - [ ] `Channel Info` > `Add People` 
  - [ ] `Channel Nav Header` > `...` > `Add People`
- [ ] `Go Back` button when no members left to add also needs to know which screen to `go back` to

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47655

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note

```

enahum commented on app/actions/remote/user.ts:
```diff
@@ -530,6 +530,46 @@ export const fetchProfilesInTeam = async (serverUrl: string, teamId: string, pag
+    perPage: number = General.PROFILE_CHUNK_SIZE,
+    fetchOnly = false,
+) => {
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return {error: `${serverUrl} database not found`};
+    }
```
same as in the other PR

enahum commented on app/screens/channel_add_people/channel_add_people.tsx:
```diff
@@ -0,0 +1,324 @@
+                    value={term}
+                />
+            </View>
+            {/*
+                https://mattermost.atlassian.net/browse/MM-48489
+                V1 does not have the selected users modal.
+                Add this back in after build the scrollable selectable users panel
+            */}
+            {/* {selectedCount > 0 &&
+                 <SelectedUsers
+                     selectedIds={selectedIds}
+                     warnCount={General.MAX_ADD_USERS - 2}
+                     maxCount={General.MAX_ADD_USERS}
+                     onRemove={handleRemoveProfile}
+                     teammateNameDisplay={teammateNameDisplay}
+                 />
+             } */}
```
???

jfrerich replied:
selected users panel added and now being use in this PR.

enahum commented on app/screens/channel_add_people/channel_add_people.tsx:
```diff
@@ -0,0 +1,324 @@
+import {Keyboard, Platform, StyleSheet, View} from 'react-native';
+import {SafeAreaView} from 'react-native-safe-area-context';
+
+// import SelectedUsers from '@components/selected_users_panel';
```
??

enahum commented on app/screens/channel_add_people/channel_add_people.tsx:
```diff
@@ -0,0 +1,324 @@
+// import SelectedUsers from '@components/selected_users_panel';
+import {addMembersToChannel} from '@actions/remote/channel';
+import {fetchProfilesNotInChannel, searchProfiles} from '@actions/remote/user';
+import useDidUpdate from '@app/hooks/did_update';
```
```suggestion
import useDidUpdate from '@hooks/did_update';
```

matthewbirtch reviewed (changes requested):
@jfrerich a few things from me:

1. Since we're not implementing this with the selected users bottom panel for now (as you described in the PR description), we should use the chips as we do in the new DM modal for now until the bottom panel is implemented. Otherwise you won't be able to see who you've selected as you scroll through a long list
2. As I type in the search input, it immediately shows an empty state at first and then the loader displays. Can we avoid displaying the empty state until we're certain there are no results? This may apply to other search scenarios in the app as well. I feel like I've called this out before.
3. In a channel where all team members are already in the channel, there are no members to add, so I guess we'll need a slightly different empty state to show here. In this case we should probably hide the search input as well.

<img width="564" alt="image" src="https://user-images.githubusercontent.com/2040554/202763290-8ea0454e-3e88-4d04-b9da-fc4a2ced8ffa.png">

We can just do something like this ([figma link here](https://www.figma.com/file/uCS4dYCHSS5NO7PjlVUBzI/MM-19947-Mobile-Multi-select?node-id=3154%3A96661&t=XvFpID6yqJVD8O8I-4) as well): 

<img src="https://user-images.githubusercontent.com/2040554/202764559-56b4e94b-7ae8-44cb-881d-67cba155cf49.png" width="400">


4. In the design, a toast was supposed to display when you go back to the main channel info screen
<img src="https://user-images.githubusercontent.com/2040554/202764920-ca3f63a0-ebc3-4a8b-a98c-3d8c695638c1.png" width="400">

5. If we can go the toast approach, I'm hoping we won't need this loading state (see below)? If we do still need this loading state, I would center the loader on the screen instead of where it is positioned right now.
<img width="400" alt="image" src="https://user-images.githubusercontent.com/2040554/202765882-0d8df56b-d5df-4c57-abb8-bf28a4bcac59.png">

6. I can't seem to consistently reproduce this, but there were a few times during my testing where the `>` back button disappeared from the modal's top bar so I couldn't go back to the info screen. In this scenario, if I hit the `Add` button, the screen hung up on the loading state and never finished and 'Possible unhandled promise' error displayed in the simulator. I had to reload the entire app to get out of this state.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/2040554/202766502-78aefae2-c1aa-44d3-b4ef-67b69cb25f86.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/2040554/202766786-a38afed3-d279-4085-a8d0-56a05323f01e.png">

7. On the channel info screen, tapping 'Members' resulted in this 'Add members' screen. This should not be the case. I know you're implementing the view/manage members screen separately though, so this may not be an issue.

jfrerich commented:
Here's an update on your response, @matthewbirtch 

> 1. Since we're not implementing this with the selected users bottom panel for now (as you described in the PR description), we should use the chips as we do in the new DM modal for now until the bottom panel is implemented. Otherwise you won't be able to see who you've selected as you scroll through a long list

**SUGGESTION** The issue with the selected users and why it probably wasn't used in V1 is because it isn't scrollable.  I propose having it added after https://github.com/mattermost/mattermost-mobile/pull/6699 is merged.  I have added a ticket to do this after merging that PR

> 2. As I type in the search input, it immediately shows an empty state at first and then the loader displays. Can we avoid displaying the empty state until we're certain there are no results? This may apply to other search scenarios in the app as well. I feel like I've called this out before.

**FIXED** for `AddMembers` and `CreateDirectMessage` screens.  The loading spinner shows instead of the emtpy state.

> 3. In a channel where all team members are already in the channel, there are no members to add, so I guess we'll need a slightly different empty state to show here. In this case we should probably hide the search input as well.

**FIXED** I added the screen you designed. with the back button.

> 4. In the design, a toast was supposed to display when you go back to the main channel info screen

**IN PROGRESS**

> 5. If we can go the toast approach, I'm hoping we won't need this loading state (see below)? If we do still need this loading state, I would center the loader on the screen instead of where it is positioned right now.

**IN PROGRESS**

> 6. I can't seem to consistently reproduce this, but there were a few times during my testing where the > back button disappeared from the modal's top bar so I couldn't go back to the info screen. In this scenario, if I hit the Add button, the screen hung up on the loading state and never finished and 'Possible unhandled promise' error displayed in the simulator. I had to reload the entire app to get out of this state.

**IN PROGRESS** I am seeing this but don't have a clear set of repro steps yet.  I'll continue working on this one. 

> 7. On the channel info screen, tapping 'Members' resulted in this 'Add members' screen. This should not be the case. I know you're implementing the view/manage members screen separately though, so this may not be an issue.

**CORRECT** that screen will be implemented via https://github.com/mattermost/mattermost-mobile/pull/6771

jfrerich commented on assets/base/i18n/en.json:
```diff
@@ -853,6 +859,8 @@
   "share_feedback.title": "Would you share your feedback?",
   "smobile.search.recent_title": "Recent searches in {teamName}",
   "snack.bar.favorited.channel": "This channel was favorited",
+  "snack.bar.channel.member.added": "1 member added",
```
use intl `plural` option and remove second id  https://formatjs.io/docs/intl-messageformat/

avinashlng1080 reviewed (approved):
LGTM. Approving so as not to blocked PR.  However, some lines are duplicated over 1 or 2 files.  Can these duplicates be reduced?
![code duplicates](https://user-images.githubusercontent.com/43164508/214279020-5c66da32-1dec-4ee4-9bc6-417e4965a19c.png)

avinashlng1080 commented:
@jfrerich to resolve merge conflict

jfrerich commented on app/screens/channel_add_people/index.tsx:
```diff
@@ -0,0 +1,36 @@
+import type {WithDatabaseArgs} from '@typings/database/database';
+
+const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
+    const currentChannel = observeCurrentChannel(database);
```
The need the currentChannel observable for `isGroupConstrained` and to get `channelId`.  Thats why `currentChannel` is it's own variable

Willyfrog commented:
what's left on this PR to be merged?

esethna commented:
@amyblais we discussed potentially getting an exception to take the members related PRs into v2.1 given the impact on user experience and feedback in the stores. The risk can be mitigated by getting it on the beta app launch tomorrow and testing there thoroughly. 

If we're in alignment let's aim to cherry pick these into v2.1

matthewbirtch commented:
@jfrerich we still need to add the 'plus' button for adding people on the View/Manage Members screen, right? Or is that going to be done in another PR?

![image](https://user-images.githubusercontent.com/2040554/216682996-a510dc59-094f-4942-b269-8e5b3b92d9e5.png)

jfrerich commented:
@matthewbirtch, I've noticed those changes too. This was after merging latest master. I'm making updates now.

jfrerich commented:
> what's left on this PR to be merged?

@Willyfrog, I'm currently reviewing changes need after we moved the new react-native bottom sheet.  See @matthewbirtch comment above. 

I have a fix for his comment and am working on the tablet version fixes.  tapping on search in tablet is wrong after the merge.

enahum commented:
closing this PR for now
